### DISCUSSION TITLE
How to activate gpu hardware acceleration?

### DISCUSSION THREAD

**Discussion by anazei** [ImageMagick/ImageMagick#5804] [Help] [answered]
Hi Guys

I know this is probably a basic question but I just wanted to say that I'm not coming straight here to ask this question without first having searched for the answer.

I spent a considerable amount of time googling and trying to figure out how to activate the hardware acceleration for resize operations for Image Magick. I read through as many articles and questions and forums as I could including refining my search many, many times, mostly on Google and some on this discussion forum. I cannot find a straightforward answer to this. 

I was initially using the Windows x64 16 bit colour prebuilt .exe version and it's not clear to me how to enable hardware acceleration. 

Here's what I've done so far from what I've pieced together: 

- I downloaded the C++ build and followed the instructions carefully to build it from source after downloading Visual Studio and ensuring the option "OpenCL" was checked off in the build wizard tool

- I disabled the line in the source code from https://github.com/ImageMagick/ImageMagick/discussions/2545 to try to enable GPU support for Nvidia since I use the Geforce 2080.

- I read on multiple sites that Nvidia GPUs are not supported so I actually went out and bought a $300 AMD GPU today but it doesn't seem to be processing any faster than before. 

- I ran "magick -version" in the command line and it says "OpenCL" is supported in its features list

- I edited the file called "configuration.xml" and added configure name="MAGICK_OCL_DEVICE" value="true" though I'm not entirely sure this is correct

- I ran "set MAGICK_OCL_DEVICE=true" in the command line before running magick mogrify -path c:\filepath -resize "1280x720>" *.jpeg

All this actually somehow made it worse. 
The mogrify command would just fail after 50 minutes and zero image processing was done. 
I ran three tests with variations in the magick command and I confirmed something was wrong because it failed three times.
I ended up uninstalling Image Magick and reinstalling it and it worked again with factory settings.

The version I built from source carefully following the instructions was even worse. 
On two different computers that I tested, not only does it not use hardware acceleration but it somehow runs up to 30% slower than the prebuilt .exe version!

When I check the task manager, the AMD GPU is 0% and the fans aren't even spinning which means it's probably not doing any heavy lifting work.

- I ran "set MAGICK_OCL_DEVICE=true" in the command line before running magick mogrify -path c:\filepath -resize "1280x720>" *.jpeg

All this actually somehow made it worse. 
The mogrify command would just fail after 50 minutes and zero image processing was done. 
I ran three tests with variations in the magick command and I confirmed something was wrong because it failed three times.
I ended up uninstalling Image Magick and reinstalling it and it worked again with factory settings.

The version I built from source carefully following the instructions was even worse. 
On two different computers that I tested, not only does it not use hardware acceleration but it somehow runs up to 30% slower!

When I check the task manager, the AMD GPU is 0% and the fans aren't even spinning which means it's probably not doing any heavy lifting work.

I am asking this question because I'm pretty much at wit's end. 
Can you guys provide some instructions on how to enable hardware acceleration in Image Magick and some way to confirm hardware acceleration is working?

**Comment by anazei** [comment, marked as answer]
Okay, I ended up swapping the AMD into my much faster main computer and I ran two tests. 

It looks like it works on this computer but was super slow on my older computer.

I'm just running the Windows prebuilt executable version with no modifications to the configuration.xml or anything else.

I ran one test with the flag "set MAGICK_OCL_DEVICE=true" in command prompt before running the magick command and one test with no flag turned on. 
The original computer the AMD was being run on is almost 10 years old so that might have severely limited the GPU's capabilities.
On the older original computer, it seem to make zero difference whether hardware acceleration flag was turned on or not because in multiple tests with the exact same set of data, it took the same amount of time to process.

With the AMD plugged into the newer computer:
No hardware acceleration took 23 minutes to process 4.5gb of images of 30,602 jpegs.
Hardware acceleration took 18 minutes to process the exact same 4.5gb of images.
It looked like the GPU was only using 4% of its potential according to Windows Task Manager.

AMD's 4% hardware acceleration is about 22% faster than just raw CPU (I'm using the Intel 10920K).
What I'm hoping to do is to scale up this 22% advantage by running 24 instances (theoretical max of my 12 core CPU) of Image Magick simultaneously and hope the GPU shines through its ability to run all 24 instances with 4% GPU potential in each instance.

**Comment by snibgo** [comment]
I know almost nothing about OpenCL.

If we add "-debug all", we get some diagnostics. For example:
```
magick -debug all toes.png -resize 50% NULL: 2>&1 |grep opencl

2022-11-25T14:36:48+00:00 0:00.022 0.000u 7.1.0 Accelerate magick[11384]: opencl.c/GetOpenCLCacheDirectory/371/Accelerate
2022-11-25T14:36:48+00:00 0:00.089 0.000u 7.1.0 Accelerate magick[11384]: opencl.c/LoadOpenCLDevices/2427/Accelerate
  Domain: Path; rights=Read; pattern="C:\Users\Alan\AppData\Local\ImageMagick\magick_opencl_Intel(R)_Iris(R)_Xe_Graphics_1b5f6663_64.bin" ...
2022-11-25T14:36:48+00:00 0:00.091 0.000u 7.1.0 Accelerate magick[11384]: opencl.c/LoadCachedOpenCLKernels/1328/Accelerate
  Loaded cached kernels: "C:\Users\Alan\AppData\Local\ImageMagick\magick_opencl_Intel(R)_Iris(R)_Xe_Graphics_1b5f6663_64.bin"
2022-11-25T14:36:48+00:00 0:00.093 0.000u 7.1.0 Accelerate magick[11384]: opencl.c/AcquireOpenCLKernel/724/Accelerate
2022-11-25T14:36:48+00:00 0:00.093 0.000u 7.1.0 Accelerate magick[11384]: opencl.c/AcquireOpenCLKernel/724/Accelerate
```

**Comment by snibgo** [comment]
The performance improvement from OpenCL will depend heavily on machine architecture and the workload. For example, on a laptop with "Intel(R) Iris(R) Xe Graphics" GPU, 7.9 GB "Shared GPU memory", whatever that is, and this command:
```
magick AGA_1434_gms.tiff -blur 0x1000 NULL:
```
... where the input image has 4924x7378 pixels, this takes 1482 seconds without OpenCL, or 7.5 seconds with OpenCL. The improvement is 200 times, or a reduction of 99.5%.

This is using pre-built IM on Windows which includes OpenCL but (strangely) not OpenMP.

**Comment by snibgo** [comment]
Parallel processing, whether using multi-threading or simultaneous but separate processes or GPU, is a complex business. To get best overall performance (eg "total frames processed per hour") we generally aim to saturate all resources equally: CPU, GPU, memory, disk access, and the IO channels between those components. If one resource is used less than the others, there is opportunity to increase the performance. This might need tweaking of parameters, or redesigning the solution architecture.

Your approach is good: try different solutions for the work you need to do, and measure the differences. Beware of accidents. For example, if your images are large then multiple instances of IM may exhaust physical memory, so IM will cache pixels to disk, which is much slower especially if not using SDD.
